notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala

```diff
@@ -110,6 +109,19 @@ case class LocationIO(configuration: Configuration, s3Client: AmazonS3Client) {
       }}.as(state)
     }
 
+  /** read a file as a stream of bytes and compute some state value */
+  def streamBytes[A](location: Location, empty: => A, bufferSize: Int = 4096)(f: (Buffer, A) => A): RIO[A] =
+    RIO.io(empty).flatMap { s =>
+      var state = s
+      readUnsafe(location) { in => RIO.io {
+        val buffer = Buffer.wrapArray(Array.ofDim[Byte](bufferSize), 0, bufferSize)
+        var length = 0
+        while ({ length = in.read(buffer.bytes, buffer.offset, bufferSize); length != -1 }) {
+          state = f(Buffer.allocate(buffer, length), state)
```

Comment:
We need to work on these buffer idioms, this is safe in this context, but not obviously so, and could easily break if we are not careful. The specific issue is buffer can return a new buffer and that case is never handled, this should really be:

```
buffer = Buffer.allocate(buffer, length)
state = f(buffer, state)
```
Alternatively, this really doesn't need to be allocate, there could just be "setLength" on Buffer that is guaranteed to re-use the same buffer or fail.

Thoughts @charleso ?